scripts/tools/ResourceIndexGarbageCollector.php

```diff
@@ -0,0 +1,199 @@
+                'description' => 'Search offset',
+                'defaultValue' => 0,
+            ],
+            'index' => [
+                'prefix' => 'i',
+                'longPrefix' => 'index',
+                'flag' => false,
+                'description' => 'The index to clean',
+                'required' => true,
+            ],
+        ];
+    }
+
+    protected function provideDescription(): string
+    {
+        return 'Clear invalid indexes';
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function run(): Report
+    {
+        $index = $this->getOption('index');
+        $limit = (int)$this->getOption('limit');
+        $offset = (int)$this->getOption('offset');
+        $report = Report::createInfo(sprintf('Cleaning documents for index "%s"', $index));
+        $searcher = $this->getSearch();
+        $totalRemoved = 0;
+        $urisToRemove = [];
+
+        do {
+            $removed = $this->incrementUrisToRemove($index, $offset, $limit, $urisToRemove);
+            $offset += $limit;
+        } while ($removed !== null);
+
+        foreach ($urisToRemove as $uri) {
+            if ($searcher->remove($uri)) {
+                $totalRemoved++;
+
+                continue;
+            }
+
+            $errorMessage = sprintf(
+                'Error removing resource "%s" for index "%s" (offset:%s, limit:%s)',
+                $uri,
+                $index,
+                $offset,
+                $limit
+            );
+
+            $report->add(Report::createError($errorMessage));
+
+            $this->logError($errorMessage);
+        }
+
+        $message = sprintf(
+            'Total of %s document(s) removed from index %s',
+            $totalRemoved,
+            $index
+        );
+
+        $report->add(Report::createSuccess($message));
+
+        $this->logInfo($message);
+
+        return $report;
+    }
+
+    private function incrementUrisToRemove(string $index, int $offset, int $limit, array &$urisToRemove): ?array
+    {
+        $results = $this->search($index, $offset, $limit);
+
+        if ($results->count() === 0) {
+            return null;
+        }
+
+        foreach ($results as $result) {
+            $resource = $this->getResource($result['id']);
+
+            if ($this->isMissingResource($resource)) {
+                $urisToRemove[] = $result['id'];
+            }
+        }
+
+        return $urisToRemove;
+    }
+
+    private function isMissingResource(core_kernel_classes_Resource $resource): bool
+    {
+        if ($this->isMissingType($resource)) {
+            return true;
+        }
+
+        if ($resource->isClass()) {
+            return false;
+        }
+
+        if ($resource->exists()) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private function isMissingType(core_kernel_classes_Resource $resource): bool
+    {
+        foreach ($resource->getTypes() as $class) {
+            if (!$class->exists()) {
```

Comment:
Under what circumstances that might work?

Reply:
Here: https://oat-sa.atlassian.net/browse/ADF-569?focusedCommentId=144690

The reason I put it is because we have real cases where the class was deleted, but the reference with the resource type remains (invalid).